### PR TITLE
Dataflow: Refactor FlowSummaryImpl to synthesize nodes independently from DataFlow::Node.

### PR DESCRIPTION
**Pull request by aschackmull**
The `FlowSummaryImpl` library is thoroughly entangled with various bits and pieces from the language-specific parts. This PR does some untangling by making it possible to synthesize summary nodes and `DataFlowCall`s in summarized callables without depending on `DataFlow::Node`.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll [hunk 1]
```diff
@@ -447,61 +441,61 @@ module Private {
...
 
     /** Holds if this summary node is an out node for `call`. */
-    predicate isOut(DataFlowCall call) { FlowSummaryImpl::Private::summaryOutNode(call, this, _) }
-  }
-
-  SummaryNode getSummaryNode(SummarizedCallable c, FlowSummaryImpl::Private::SummaryNodeState state) {
-    result = TSummaryInternalNode(c, state)
+    predicate isOut(DataFlowCall call) {
+      FlowSummaryImpl::Private::summaryOutNode(call, this.getSummaryNode(), _)
+    }
```
Comment (yoff) [inline java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll:480]: I wonder why you chose not to add a field `summaryNode` corresponding to `this.getSummaryNode()` and bind that in a charpred. Then we could refer to it when implementing the other predicates (here and in extending classes). Is it just a matter of style or are there other concerns?

File: java/ql/lib/semmle/code/java/dataflow/internal/FlowSummaryImpl.qll [hunk 2]
```diff
@@ -859,27 +904,27 @@ module Private {
...
     pragma[noinline]
     private predicate viableParam(
-      DataFlowCall call, SummarizedCallable sc, ParameterPosition ppos, ParamNode p
+      DataFlowCall call, SummarizedCallable sc, ParameterPosition ppos, SummaryParamNode p
     ) {
       exists(DataFlowCallable c |
         c = inject(sc) and
-        p.isParameterOf(c, ppos) and
+        p = TSummaryParameterNode(sc, ppos) and
         c = viableCallable(call)
```
Comment (yoff) [inline java/ql/lib/semmle/code/java/dataflow/internal/FlowSummaryImpl.qll:922]: This is morally just
```
call = sc.getACall() and
p = TSummaryParameterNode(sc, ppos)
```
so we might be able to get rid of `c` here? In fact, it feels like we can almost get rid of `inject` altogether?

**Review by yoff — APPROVED** [review]
Just some curious questions, but this looks good from the Python side, assuming the evaluations are acceptable (it looks like they have to be re-run?).

**Review by geoffw0 — APPROVED** [review]
Swift LGTM - many of our tests are looking better now these nodes have proper locations. :+1:

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll (hunk 1 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll:480]: There are unfortunately other concerns. The optimiser still has some shortcomings wrt. fields, so adding an equivalent getter instead of using a charpred is supposedly better at the moment. That way the optimiser is able to infer that `FlowSummaryNode` is equal to `TFlowSummaryNode` at an earlier stage (a charpred can in general break this assumption).

File: go/ql/lib/semmle/go/dataflow/internal/FlowSummaryImplSpecific.qll [hunk 3]
```diff
@@ -17,7 +17,7 @@ private module FlowSummaries {
 
 class SummarizedCallableBase = Callable;
 
-DataFlowCallable inject(SummarizedCallable c) { result.asSummarizedCallable() = c }
+DataFlowCallable inject(SummarizedCallable c) { result.asSummarizedCallable() = c or none() }
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply go/ql/lib/semmle/go/dataflow/internal/FlowSummaryImplSpecific.qll:20]: It makes things compile by hiding a fact from the compiler 🙈 
It's the non-trivial emptiness check that fails - that's disabled for things in parameterised modules for good reason, but since this is still a _py_-rameterised module we still have the check enabled.

**Comment by aschackmull** [comment]
> Also, due to Python not being totally perfect about keeping things private, we get [a QLDoc failure](https://github.com/github/semmle-code/actions/runs/5080532357/jobs/9257070912#step:23:35).

Yep. I intend to ignore that.

**Review by owen-mc — APPROVED** [review]
LGTM for go

File: java/ql/lib/semmle/code/java/dataflow/internal/FlowSummaryImpl.qll (hunk 2 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/internal/FlowSummaryImpl.qll:922]: That sounds right, and certainly looks like a step in the right direction in terms of further disentangling things. We just have to be sure that `getACall` is part of the implicit language-specific api.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll (hunk 1 above)
Comment (yoff) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll:480]: I see, thanks for the explanation.

**Review by owen-mc — DISMISSED** [review]

**Comment by yoff** [comment]
The failing Python Language Tests should be solved by a rebase (we recently made the parser more robust, resulting in this change).

**Comment by yoff** [comment]
Which source suite did you use for the DCA experiment for Python? We just disabled `piipan` from `nightly` because the repo has been deleted. The other error [looks like an actions thing..](https://github.com/dsp-testing/codeql-dca-worker_python/actions/runs/5222618795/jobs/9428340986#step:8:18)

**Comment by aschackmull** [comment]
> Which source suite did you use for the DCA experiment for Python? We just disabled `piipan` from `nightly` because the repo has been deleted. The other error [looks like an actions thing..](https://github.com/dsp-testing/codeql-dca-worker_python/actions/runs/5222618795/jobs/9428340986#step:8:18)

I may have failed to update dca before launching...

File: csharp/ql/lib/ext/System.Collections.Generic.model.yml [hunk 4]
```diff
@@ -40,18 +40,18 @@ extensions:
...
-      - ["System.Collections.Generic", "List<>", False, "GetRange", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Generic", "List<>", False, "GetRange", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
       - ["System.Collections.Generic", "List<>", False, "InsertRange", "(System.Int32,System.Collections.Generic.IEnumerable<T>)", "", "Argument[1].Element", "Argument[this].Element", "value", "manual"]
-      - ["System.Collections.Generic", "List<>", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections.Generic", "List<>", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Generic", "List<>", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Generic", "List<>", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.Generic.model.yml:54]: Same comment as above.

File: csharp/ql/lib/ext/System.Collections.Generic.model.yml [hunk 5]
```diff
@@ -40,18 +40,18 @@ extensions:
...
-      - ["System.Collections.Generic", "List<>", False, "GetRange", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Generic", "List<>", False, "GetRange", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
       - ["System.Collections.Generic", "List<>", False, "InsertRange", "(System.Int32,System.Collections.Generic.IEnumerable<T>)", "", "Argument[1].Element", "Argument[this].Element", "value", "manual"]
-      - ["System.Collections.Generic", "List<>", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections.Generic", "List<>", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Generic", "List<>", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply csharp/ql/lib/ext/System.Collections.Generic.model.yml:53]: Right, of course.

File: csharp/ql/lib/ext/System.Collections.Generic.model.yml [hunk 6]
```diff
@@ -79,7 +79,7 @@ extensions:
       - ["System.Collections.Generic", "SortedList<,>+KeyList", False, "Clear", "()", "", "Argument[this].WithoutElement", "Argument[this]", "value", "manual"]
       - ["System.Collections.Generic", "SortedList<,>+ValueList", False, "Clear", "()", "", "Argument[this].WithoutElement", "Argument[this]", "value", "manual"]
       - ["System.Collections.Generic", "SortedSet<>", False, "GetEnumerator", "()", "", "Argument[this].Element", "ReturnValue.Property[System.Collections.Generic.SortedSet<>+Enumerator.Current]", "value", "manual"]
-      - ["System.Collections.Generic", "SortedSet<>", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Generic", "SortedSet<>", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.Generic.model.yml:82]: This one looks good (Reverse is not side effect based in the SortedSet<> class)

File: csharp/ql/lib/ext/System.Collections.Immutable.model.yml [hunk 7]
```diff
@@ -20,7 +20,7 @@ extensions:
       - ["System.Collections.Immutable", "ImmutableArray<>+Builder", False, "AddRange<>", "(System.Collections.Immutable.ImmutableArray<TDerived>+Builder)", "", "Argument[0].Element", "Argument[this].Element", "value", "manual"]
       - ["System.Collections.Immutable", "ImmutableArray<>+Builder", False, "AddRange<>", "(TDerived[])", "", "Argument[0].Element", "Argument[this].Element", "value", "manual"]
       - ["System.Collections.Immutable", "ImmutableArray<>+Builder", False, "GetEnumerator", "()", "", "Argument[this].Element", "ReturnValue.Property[System.Collections.Generic.IEnumerator<>.Current]", "value", "manual"]
-      - ["System.Collections.Immutable", "ImmutableArray<>+Builder", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Immutable", "ImmutableArray<>+Builder", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.Immutable.model.yml:23]: Remove.

File: csharp/ql/lib/ext/System.Collections.Generic.model.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aschackmull) [reply csharp/ql/lib/ext/System.Collections.Generic.model.yml:53]: Ah, I see it's not generally the case - some of them return a modified collection and some work by side effect.

File: csharp/ql/lib/ext/System.Collections.Immutable.model.yml (hunk 7 above)
Comment (aschackmull) [reply csharp/ql/lib/ext/System.Collections.Immutable.model.yml:23]: Are you sure? That's a builder, so I would assume that it returns a modified builder.

File: csharp/ql/lib/ext/System.Collections.Immutable.model.yml [hunk 8]
```diff
@@ -63,10 +63,10 @@ extensions:
...
-      - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "GetRange", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "GetRange", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
       - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "InsertRange", "(System.Int32,System.Collections.Generic.IEnumerable<T>)", "", "Argument[1].Element", "Argument[this].Element", "value", "manual"]
-      - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.Immutable.model.yml:68]: Remove.

File: csharp/ql/lib/ext/System.Collections.Immutable.model.yml [hunk 9]
```diff
@@ -63,10 +63,10 @@ extensions:
...
+      - ["System.Collections.Immutable", "ImmutableList<>+Builder", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.Immutable.model.yml:69]: Remove.

File: csharp/ql/lib/ext/System.Collections.model.yml [hunk 10]
```diff
@@ -4,17 +4,17 @@ extensions:
...
       - ["System.Collections", "ArrayList", False, "InsertRange", "(System.Int32,System.Collections.ICollection)", "", "Argument[1].Element", "Argument[this].Element", "value", "manual"]
       - ["System.Collections", "ArrayList", False, "Repeat", "(System.Object,System.Int32)", "", "Argument[0]", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections", "ArrayList", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections", "ArrayList", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections", "BitArray", False, "Clone", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Collections", "Hashtable", False, "Clone", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Collections", "ArrayList", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.model.yml:14]: Remove

File: csharp/ql/lib/ext/System.Collections.model.yml [hunk 11]
```diff
@@ -4,17 +4,17 @@ extensions:
...
+      - ["System.Collections", "ArrayList", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.model.yml:15]: Remove

File: csharp/ql/lib/ext/System.Runtime.CompilerServices.model.yml [hunk 12]
```diff
@@ -6,6 +6,6 @@ extensions:
       - ["System.Runtime.CompilerServices", "ConditionalWeakTable<,>", False, "Clear", "()", "", "Argument[this].WithoutElement", "Argument[this]", "value", "manual"]
       - ["System.Runtime.CompilerServices", "ConfiguredTaskAwaitable<>", False, "GetAwaiter", "()", "", "Argument[this].SyntheticField[m_configuredTaskAwaiter]", "ReturnValue", "value", "manual"]
       - ["System.Runtime.CompilerServices", "ConfiguredTaskAwaitable<>+ConfiguredTaskAwaiter", False, "GetResult", "()", "", "Argument[this].SyntheticField[m_task_configured_task_awaitable].Property[System.Threading.Tasks.Task<>.Result]", "ReturnValue", "value", "manual"]
-      - ["System.Runtime.CompilerServices", "ReadOnlyCollectionBuilder<>", False, "Reverse", "()", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
-      - ["System.Runtime.CompilerServices", "ReadOnlyCollectionBuilder<>", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[0].Element", "ReturnValue.Element", "value", "manual"]
+      - ["System.Runtime.CompilerServices", "ReadOnlyCollectionBuilder<>", False, "Reverse", "()", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Runtime.CompilerServices.model.yml:9]: Remove

File: csharp/ql/lib/ext/System.Runtime.CompilerServices.model.yml [hunk 13]
```diff
@@ -6,6 +6,6 @@ extensions:
...
+      - ["System.Runtime.CompilerServices", "ReadOnlyCollectionBuilder<>", False, "Reverse", "(System.Int32,System.Int32)", "", "Argument[this].Element", "ReturnValue.Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Runtime.CompilerServices.model.yml:10]: Remove

File: csharp/ql/lib/ext/System.Collections.Immutable.model.yml (hunk 7 above)
Comment (michaelnebel) [reply csharp/ql/lib/ext/System.Collections.Immutable.model.yml:23]: Yes, that was my guess as well before looking into the documentation: https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.builder.reverse?view=net-7.0

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 14]
```diff
@@ -1609,37 +1592,40 @@ private module OutNodes {
     }
   }
 
-  private class SummaryOutNode extends SummaryNode, OutNode {
-    SummaryOutNode() { FlowSummaryImpl::Private::summaryOutNode(_, this, _) }
+  private class SummaryOutNode extends FlowSummaryNode, OutNode {
+    SummaryOutNode() { FlowSummaryImpl::Private::summaryOutNode(_, this.getSummaryNode(), _) }
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1596]: Same about fields.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 15]
```diff
@@ -2130,14 +2119,15 @@ private module PostUpdateNodes {
     override string toStringImpl() { result = "[post] " + cfn.toString() }
   }
 
-  private class SummaryPostUpdateNode extends SummaryNode, PostUpdateNode {
+  private class SummaryPostUpdateNode extends FlowSummaryNode, PostUpdateNode {
     SummaryPostUpdateNode() {
-      FlowSummaryImpl::Private::summaryPostUpdateNode(this, _) and
+      FlowSummaryImpl::Private::summaryPostUpdateNode(this.getSummaryNode(), _) and
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:2124]: Field.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll [hunk 16]
```diff
@@ -505,6 +505,9 @@ module Private {
     or
     // Add the post-update node corresponding to the requested argument node
     outputState(c, s) and isCallbackParameter(s)
+    or
+    // Add the parameter node for parameter side-effects
+    outputState(c, s) and s = SummaryComponentStack::argument(_)
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll:510]: Why was this addition needed?

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll [hunk 17]
```diff
@@ -559,6 +562,42 @@ module Private {
...
 
+  private newtype TSummaryNode =
+    TSummaryInternalNode(SummarizedCallable c, SummaryNodeState state) {
+      summaryNodeRange(c, state)
+    } or
+    TSummaryParameterNode(SummarizedCallable c, ParameterPosition pos) {
+      summaryParameterNodeRange(c, pos)
+    }
+
+  class SummaryNode extends TSummaryNode {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll:573]: abstract

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 18]
```diff
@@ -757,47 +750,43 @@ private module ParameterNodes {
   }
 
   /** A parameter for a library callable with a flow summary. */
-  class SummaryParameterNode extends ParameterNodeImpl, TSummaryParameterNode {
-    private FlowSummaryImpl::Public::SummarizedCallable sc;
-    private ParameterPosition pos_;
+  class SummaryParameterNode extends ParameterNodeImpl, FlowSummaryNode {
+    SummaryParameterNode() {
+      FlowSummaryImpl::Private::summaryParameterNode(this.getSummaryNode(), _)
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:755]: field

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 19]
```diff
@@ -857,15 +846,17 @@ private module ArgumentNodes {
     }
   }
 
-  private class SummaryArgumentNode extends SummaryNode, ArgumentNode {
-    SummaryArgumentNode() { FlowSummaryImpl::Private::summaryArgumentNode(_, this, _) }
+  private class SummaryArgumentNode extends FlowSummaryNode, ArgumentNode {
+    SummaryArgumentNode() {
+      FlowSummaryImpl::Private::summaryArgumentNode(_, this.getSummaryNode(), _)
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:851]: fields

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 20]
```diff
@@ -1071,11 +1062,11 @@ private module OutNodes {
     }
   }
 
-  private class SummaryOutNode extends SummaryNode, OutNode {
-    SummaryOutNode() { FlowSummaryImpl::Private::summaryOutNode(_, this, _) }
+  private class SummaryOutNode extends FlowSummaryNode, OutNode {
+    SummaryOutNode() { FlowSummaryImpl::Private::summaryOutNode(_, this.getSummaryNode(), _) }
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:1066]: fields

**Review by hvitved — COMMENTED** [review]
Ruby/C#/Shared LGTM. Some minor comments.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll (hunk 16 above)
Comment (aschackmull) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll:510]: Previously we were relying on a separate synthesis of parameter nodes, so when we only had a summary reference to the postupdate of a given parameter, we would now fail to synthesise the pre-update without this.

**Review by hvitved — APPROVED** [review]
I suggest doing a final DCA run before merging.
